### ISSUE TITLE
Play webm in Damus

### ISSUE THREAD

**Issue by alltheseas** [damus-io/damus#2090]
## current behavior

.webm either opens a browser (safari), or prompts a download link to .webm files (brave) on mobile

On MacOS .webm opens default browser.

## what I expect to happen

embed webm, so I can play the video in Damus without leaving the app

**Comment by alltheseas** [comment]
recreated by @tyiu

**Comment by alltheseas** [comment]
https://damus.io/nevent1qqsw0n0x5sjlve4hhrp57g2ljaqzw59edym9x943rzylhu5ycl3y66qppamhxue69uhkummnw3ezumt0d5q3gamnwvaz7tmjv4kxz7fwdehhxarj9e3xwqg5waehxw309ahx7um5wghx77r5wghxgetkqy28wumn8ghj7un9d3shjtnyv9kh2uewd9hsl66670

**Comment by alltheseas** [comment]
@ericholguin ser 👀

**Comment by alltheseas** [comment]
Webm is link, and not embedded on 1.10

https://damus.io/nevent1qqs8805y3per5fe4mpdfrp0wwnlc9gs6llqz7qvle3r7y3pr6qgy0fqkhedcs

**Comment by alltheseas** [comment]
https://damus.io/nevent1qqsfhgjy5guez4r0te40c8u2v5f8txncvaqlanjz72ahszen786qtzqpzpmhxue69uhkummnw3ezuamfdejszxrhwden5te0wfjkccte9eekummjwsh8xmmrd9skcqgdwaehxw309ahx7uewd3hkcqg5waehxw309aex2mrp0yhxgctdw4eju6t0gytmrd

@scoder1747 👀

**Comment by alltheseas** [comment]
![image](https://github.com/user-attachments/assets/bedab3d8-e32e-4e37-8c42-94610c105a44)

**Comment by alltheseas** [comment]
via Kagi LLM:

Based on the available information, here's the status of WebM playback support on Apple platforms and available solutions:

Native Support:
- **macOS**: Supports WebM streaming playback
  - Since macOS 11.3: VP8/VP9 video and Vorbis audio [^1]
  - Since macOS 12: Added Opus audio support [^1]
- **iOS**: No native WebM support in standard frameworks [^2]

Third-party Solutions:
1. **OGVKit**:
   - Lightweight iOS player widget
   - Supports WebM VP8/VP9 codecs [^3][^4]

2. **VLCKit**:
   - Can handle WebM playback
   - Implementation available through third-party solutions [^5]

Alternative Approaches:
1. **Server-side Conversion**:
   - Convert WebM to MP4 using backend services
   - Use FFmpeg for conversion [^6]

2. **Native Format**:
   - WebM is not compatible with default iOS video players like MPMoviePlayerController [^7]

For iOS app development, you'll need to either:
1. Use third-party frameworks like OGVKit or VLCKit
2. Convert WebM to iOS-compatible formats server-side
3. Provide alternative video formats for iOS devices

[^1]: [Develop advanced web content - WWDC21 - Videos - Apple Developer](https://developer.apple.com/videos/play/wwdc2021/10030/#:~:text=To%20make,Opus%20audio.)
[^2]: [Webm video format doesn't work on safari (ios) - Stack Overflow](https://stackoverflow.com/questions/72725245/webm-video-format-doesnt-work-on-safari-ios#:~:text=Webm%20format,webm%20file)
[^3]: [bvibber/OGVKit: Ogg and WebM media playback for iOS - GitHub](https://github.com/bvibber/OGVKit#:~:text=Lightweight%20iOS,widget%20%28OGVPlayerView%29)
[^4]: [GitHub - brion/OGVKit: Ogg and WebM media playback for iOS](https://github.com/brion/OgvKit#:~:text=Lightweight%20iOS,VP8/VP9%20codecs.)
[^5]: [Play .webm files in video player in iOS swift using VLCKit - GitHub](https://github.com/kuku93/webM_Player#:~:text=webM_Player.%20Play,terminal%3B%20change)
[^6]: [How to play webm file in iOS video player - Stack Overflow](https://stackoverflow.com/questions/27269075/how-to-play-webm-file-in-ios-video-player#:~:text=Programmatically%20you,use%20fluent-ffmpeg.)
[^7]: [How can I play webm streaming video in an MPMoviePlayerController?](https://stackoverflow.com/questions/24522645/how-can-i-play-webm-streaming-video-in-an-mpmovieplayercontroller#:~:text=WebM%20format,using%20MPMoviePlayerController.)

**Comment by alltheseas** [comment]
@fishcakeday any recommendations on iOS library for webm?

**Comment by alltheseas** [comment]
> https://github.com/kingslay/FFmpegKit

@danieldaquino

**Comment by SanjaySiddharth** [comment]
Hello @alltheseas , @jb55  , can we use cocoa pods to get VLCKit ?

**Comment by alltheseas** [comment]
@fishcakeday is the media expert, and he suggested https://github.com/kingslay/FFmpegKit

**Comment by danieldaquino** [comment]
> Hello [@alltheseas](https://github.com/alltheseas) , [@jb55](https://github.com/jb55) and [@danieldaquino](https://github.com/danieldaquino) can we use cocoa pods for VLCKit ? Or is there any other method on your plan ?

I would avoid CocoaPods in this project because we are already using the Swift Package Manager, and using CocoaPods would likely introduce complexities to the project configuration.

**Comment by SanjaySiddharth** [comment]
Okay. I have added this package "https://github.com/kingslay/FFmpegKit" to the app through the package manager. 

Let me know which approach is viable in this scenario :
1. Converting the .webm file to .mp4 and storing it in a temporary location.
2. Configuring the player to play .webm files/url directly 

I would also need help when I stumble at some point :) . Please correct me if I'm wrong , as of now , the main files which are responsible for the video playing capabilities are :
1. DamusVideoPlayer.swift -> Core Video player code , where most of the changes will have to be performed
2. NoteContent.swift -> changing the classification of urls based on format should be modified.

Please let me know your thoughts

**Comment by SanjaySiddharth** [comment]
Hello @fishcakeday , I am facing some issues with respect to the package you have suggested . Could you please help me as the documentation in the repository is vague ?

**Comment by danieldaquino** [comment]
> Okay. I have added this package "https://github.com/kingslay/FFmpegKit" to the app through the package manager.
> 
> Let me know which approach is viable in this scenario :
> 
> 1. Converting the .webm file to .mp4 and storing it in a temporary location.
> 2. Configuring the player to play .webm files/url directly

Hi @SanjaySiddharth! Option 2 would be the way to go. Option 1 would not be desired because converting videos could take several seconds or minutes making it not suitable for interactive app usage, and it also takes significant computing resources that could affect battery life negatively.

> I would also need help when I stumble at some point :) . Please correct me if I'm wrong , as of now , the main files which are responsible for the video playing capabilities are :
> 
> 1. DamusVideoPlayer.swift -> Core Video player code , where most of the changes will have to be performed
> 2. NoteContent.swift -> changing the classification of urls based on format should be modified.
> 
> Please let me know your thoughts

Yes, you are correct! `DamusVideoPlayer.swift` would likely play a central role in this ticket. It could become a universal interface which would talk to either `AVPlayer` (Apple's official video player) or the new webm-compatible player — depending on the video.

Thank you for offering to help with this ticket! I should point out that this ticket could become quite time-consuming, so please consider whether you have enough time availability before diving in — and please do not hesitate to reach out for guidance if you encounter difficulties at some point, we are here to help!

**Comment by SanjaySiddharth** [comment]
Hello @danieldaquino ,

Thanks for the heads-up ! Since we will not be converting the file , can we setup VLCKit and its associated Video Player only for running .webm files exclusively ? I would point out that I have not worked on VLCKit and this is a mere suggestion based on surfing the internet .

Please correct me if I'm wrong .

Regards,
SS

**Comment by danieldaquino** [comment]
> Hello [@danieldaquino](https://github.com/danieldaquino) ,
> 
> Thanks for the heads-up ! Since we will not be converting the file , can we setup VLCKit and its associated Video Player only for running .webm files exclusively ? I would point out that I have not worked on VLCKit and this is a mere suggestion based on surfing the internet .
> 
> Please correct me if I'm wrong .
> 
> Regards, SS

Yeah, it sounds like VLCKit might be what we are looking for. The problem is that it seems to be available via CocoaPods, which is a different package manager.

I wonder if there is a way to import VLCKit using Swift Package Manager, that would be nice!

**Comment by danieldaquino** [comment]
> Hello [@danieldaquino](https://github.com/danieldaquino) , how about this package : https://github.com/tylerjonesio/vlckit-spm BTW let me try if vlckit is available in SPM

Maybe. I don't really know that package.

**Comment by alltheseas** [comment]
https://github.com/damus-io/damus/pull/3622
